File: src/main/resources/assets/shaders/deferred/MergeReflection.fs
/*

Copyright (C) 2015-2020 Anarchy Engine Open Source Contributors (see CONTRIBUTORS.md)

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.

 */

in vec2 textureCoords;

out vec4 out_Color;

uniform sampler2D gMask;
uniform sampler2D baseTex;
uniform sampler2D ssrTex;
uniform sampler2D reflectionTex;

uniform bool useReflections;

#include variable MASK

void main(void) {
	vec3 image = texture(baseTex, textureCoords).rgb;
	vec4 mask = texture(gMask, textureCoords);
	if (MASK_COMPARE(mask.a, PBR_OBJECT)) {
		vec4 reflectionData = texture(reflectionTex, textureCoords);
		vec3 specular = reflectionData.rgb * reflectionData.a;
		if (useReflections) {
            vec4 ssrData = texture(ssrTex, textureCoords);
            specular = mix(specular, ssrData.rgb, ssrData.a);
			image += specular;
		} else {
			image += specular;
		}
	}
	out_Color.rgb = image;
	out_Color.a = 0.0;
}

File: src/main/java/engine/gl/pipeline/MergeReflection.java
/*
 *
 * Copyright (C) 2015-2020 Anarchy Engine Open Source Contributors (see CONTRIBUTORS.md)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */

package engine.gl.pipeline;

import static org.lwjgl.opengl.GL13C.GL_TEXTURE0;
import static org.lwjgl.opengl.GL13C.GL_TEXTURE1;
import static org.lwjgl.opengl.GL13C.GL_TEXTURE2;
import static org.lwjgl.opengl.GL13C.GL_TEXTURE3;

import engine.gl.DeferredPass;
import engine.gl.DeferredPipeline;
import engine.gl.RendererData;
import engine.gl.objects.Texture;
import engine.gl.pipeline.shaders.MergeReflectionShader;

public class MergeReflection extends DeferredPass<MergeReflectionShader> {

	public MergeReflection() {
		super("MergeReflection");
	}

	@Override
	protected MergeReflectionShader setupShader() {
		return new MergeReflectionShader();
	}

	@Override
	protected void setupTextures(RendererData rnd, DeferredPipeline dp, Texture[] auxTex) {
		dp.getMaskTex().active(GL_TEXTURE0);
		auxTex[1].active(GL_TEXTURE1);
		auxTex[0].active(GL_TEXTURE2);
		auxTex[2].active(GL_TEXTURE3);
	}

}
